server/src/main/scala/org/http4s/server/middleware/CORS.scala

```diff
@@ -25,36 +25,117 @@ import org.http4s.Method.OPTIONS
 import org.log4s.getLogger
 import org.typelevel.ci._
 import scala.concurrent.duration._
+import java.util.Objects
 
 /** CORS middleware config options.
   * You can give an instance of this class to the CORS middleware,
   * to specify its behavior
   */
-final case class CORSConfig(
-    anyOrigin: Boolean,
-    allowCredentials: Boolean,
-    maxAge: Long,
-    anyMethod: Boolean = true,
-    allowedOrigins: String => Boolean = _ => false,
-    allowedMethods: Option[Set[String]] = None,
-    allowedHeaders: Option[Set[String]] = Set("Content-Type", "Authorization", "*").some,
-    exposedHeaders: Option[Set[String]] = Set("*").some
-)
+
+final class CORSConfig private (
+    val anyOrigin: Boolean,
+    val allowCredentials: Boolean,
+    val maxAge: FiniteDuration,
+    val anyMethod: Boolean,
+    val allowedOrigins: String => Boolean,
+    val allowedMethods: Option[Set[Method]],
+    val allowedHeaders: Option[Set[String]],
+    val exposedHeaders: Option[Set[String]]
+) {
+
+  private def copy(
+      anyOrigin: Boolean = anyOrigin,
+      allowCredentials: Boolean = allowCredentials,
+      maxAge: FiniteDuration = maxAge,
+      anyMethod: Boolean = anyMethod,
+      allowedOrigins: String => Boolean = allowedOrigins,
+      allowedMethods: Option[Set[Method]] = allowedMethods,
+      allowedHeaders: Option[Set[String]] = allowedHeaders,
+      exposedHeaders: Option[Set[String]] = exposedHeaders
+  ) = new CORSConfig(
+    anyOrigin,
+    allowCredentials,
+    maxAge,
+    anyMethod,
+    allowedOrigins,
+    allowedMethods,
+    allowedHeaders,
+    exposedHeaders
+  )
+
+  def withAnyOrigin(anyOrigin: Boolean): CORSConfig = copy(anyOrigin = anyOrigin)
+
+  def withAllowCredentials(allowCredentials: Boolean): CORSConfig =
+    copy(allowCredentials = allowCredentials)
+
+  def withMaxAge(maxAge: FiniteDuration): CORSConfig = copy(maxAge = maxAge)
+
+  def withAnyMethod(anyMethod: Boolean): CORSConfig = copy(anyMethod = anyMethod)
+
+  def withAllowedOrigins(allowedOrigins: String => Boolean): CORSConfig =
+    copy(allowedOrigins = allowedOrigins)
+
+  def withAllowedMethods(allowedMethods: Option[Set[Method]]): CORSConfig =
+    copy(allowedMethods = allowedMethods)
+
+  def withAllowedHeaders(allowedHeaders: Option[Set[String]]): CORSConfig =
+    copy(allowedHeaders = allowedHeaders)
+
+  def withExposedHeaders(exposedHeaders: Option[Set[String]]): CORSConfig =
+    copy(exposedHeaders = exposedHeaders)
+
+  override def equals(x: Any): Boolean = x match {
+    case config: CORSConfig =>
+      anyOrigin === config.anyOrigin &&
+        allowCredentials === config.allowCredentials &&
+        maxAge === config.maxAge &&
+        anyMethod === config.anyMethod &&
+        allowedMethods === config.allowedMethods &&
+        allowedHeaders === config.allowedHeaders &&
+        exposedHeaders === config.exposedHeaders
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    Objects.hashCode(
```

Comment:
Wow, I didn't know this existed.  `MurmurHash3` is standard in Scala, and what case classes use.  But this is a bit more concise.  I think I like it!

Reply:
I searched up how to calculate hashes in Scala and found a different solution revolving around ``ScalaRunTime``, thankfully another answer showed this simple Java ``Objects#hashCode``

Reply:
I think it needs to be `Objects.hash` (the variadic thing that hashes multiple elements), not `Objects.hashCode` (takes a single object, which the Scala compiler is trying to auto-tuple).

Reply:
Whoops, I completely missed that. Nice catch, sorry about that!